### DISCUSSION TITLE
Microchip SAMD21 XPLAINED Pro does not start up

### DISCUSSION THREAD

**Discussion by SirTobias** [micropython/micropython#15886] [SAMD] [answered]
Hi,

I have this Microchip board to run micropython on it. I have build the current master branch with 
`make BOARD=SAMD21_XPLAINED_PRO` 

but have the same problem as described in issue [13790](https://github.com/orgs/micropython/discussions/13790).
When I connect the USB to the DEBUG port, the power status LED switches on and I can see the flash drive mounted. When I copy the firmware file (UF2) format, nothing happens until I plug it out and in the USB again with the mounted flash drive and the firmware vanished. When I instead power via TARGET USB, the power status LED also switches on, but nothing happens: Neither a flash drive is coming up after pressing twice the reset button nor a connection to the /dev/ttyACM port is possible.

IN MPLab X IDE I can see the board and tried to flash it with the firmware in hex format [via](https://microchip-mplab-harmony.github.io/reference_apps/apps/sam_d21_xpro/samd21_getting_started/readme.html), but the [Micropython bootup procedure](https://micropython.org/download/SAMD21_XPLAINED_PRO/) does not work.

Does anybody have an idea? Does the PA03 jumper has to be altered? It is now shortened to USB_ID.

The revision of the board is A09-2130/04

Thank you very much in advance.

UPDATE: When I connect in Debug mode, the /dev/ttyACM0 port is there, and the yellow status LED switches on when I connect via picocom or screen but nothing shows up in the terminal.

**Comment by robert-hh** [comment]
It seems that at the debug port you can only load .hex files and maybe .elf files. The debug port is NOT connected to the SAMD21 chip. It is connected to the DEBUG mcu.
I do not own this board, so I cannot verify this hint: I created a .hex file of the firmware, enclosed in the attached .zip file. You could try to load it through the debug port. If it works, REPL will be available at the target port.

[firmware.zip](https://github.com/user-attachments/files/17077667/firmware.zip)

**Comment by SirTobias** [comment]
Dear Robert, thank you for your answer. I tried your firmware hex file, but unfortunately, it did not work. Again, when I flash it via MPLab IPE and connect/switch after flashing to the TARGET USB, only the power LED is switched on, nothing else.
When I connect or leave it to the DEBUG USB port, I have /dev/ttyACM0 as a serial port, which I can connect to, but the terminal is left blank and does not react to any key input. The only difference is that as long the connection is active via picocom, the yellow status LED is activated.

I would raise this as an issue. As long as this does not work, I would not recommend this board for any usage with micropython.

**Comment by robert-hh** [comment]
It looks like an appropriate bootloader is missing. At the microchip forum there is some discussion about that, for example https://forum.microchip.com/s/topic/a5C3l000000UhUBEA0/t180270
You need a 8k sized bootloader.
You could also try to install one of the arduino bootloaders. https://github.com/arduino/ArduinoCore-samd/tree/master/bootloaders

P.S.: The SAMD21 XPLAINED Pro is indeed a little bit of overkill for using MicroPython. The Adafruit ItsyBitsy M0 is more convenient and has in addition a separate flash device for the file system.

Edit: This link looks interesting as well: https://forum.arduino.cc/t/burning-bootloader-on-to-samd21-xpro/918415

**Reply by robert-hh** [reply]
Sorry, I was wrong. You need an 8k sized bootloader.

**Reply by SirTobias** [reply]
Thank you very much for the quick reply. I am unsure why I need an 8 kByte bootloader: Does it not depend on the functionality (e.g., booting via USB/I2C, etc.)? Or does the micropython image start at a fixed RAM address? 
I am investigating these bootloaders and hope, that will work: https://github.com/Microchip-MPLAB-Harmony/bootloader_apps_usb

**Reply by robert-hh** [reply]
> I am unsure why I need an 8 kByte bootloader: 

The MicroPython code starts at Flash address 0x2000.

**Reply by robert-hh** [reply]
I got a SAMD21 XPLAINED Pro board and installed a sample bootloader from the Adafruit repositories https://github.com/adafruit/uf2-samdx1/releases. I selected the bootloader-generic-v3.16.0.bin model. I programmed that bootloader to the board. ~~You may have to convert it to Intel hex format~~.  When connected to the target port, the device then boots and shows a new drive with the name "SAMD21". Copy the firmware.uf2 file to that drive and you're done. I attached the Intel Hex version of that bootloader file.

If you want to start the bootloader again later, push Reset twice or call machine.bootloader() from the REPL prompt.

[bootloader.zip](https://github.com/user-attachments/files/17148857/bootloader.zip)

**Reply by robert-hh** [reply]
You can just copy bootloader.hex to the `XPLAINED` drive. Then it will be flashed to the board. After that, when connecting to the target USB port, you should see a `SAMD21` drive. Eventually you have to push Reset twice as one of the three methods to start the bootloader.
Edit: You can as well copy the .bin version of the bootloader to the `XPLAINED` drive. It will be installed at address 0, as it should.

**Reply by robert-hh** [reply]
Updated deploy instructions below:

The SAMD21 Xplained Pro board is shipped without a bootloader. For use
with MicroPyhton a suitable bootloader has be be installed first. The
following procedure has been found to work and be simple:

1. Get an existing bootloader. Adafruit provides a large set of suitable
bootloaders at https://github.com/adafruit/uf2-samdx1/releases. The
samd21 bootloader have  size of 8k. The file bootloader-generic-v3.16.0.bin
for instance is a suitable bootloader.
2. Convert the .bin file to Intel Hex format starting at address 0.
3. Connect your board to the debug port. A drive with the name XPLAINED
shall appear.
4. Copy the Intel hex file of the bootloader to that drive.
5. Connect your board to the target port. A drive with the name SAMD21 should
appear. If not, push reset twice. Then it should appear. If that does not
happen, the bootloader was not properly installed or is not compatible.
6. Copy the MicroPython firmware, a.uf2 file, to the SAMD21 drive. When the SAMD21
drive disappears, MicroPython is installed.

From now on only steps 5 and 6 are needed to update MicroPython. You can use the
usual methods to invoke the bootloader, namely:

- Push Reset twice.
- Call machine.bootloader().
- Use the touch 1200 procedure by switching the USB baud rate to 1200 baud and back.

At the above link above there are as well .uf2 versions of the bootloader
which one can install using steps 5. and 6. above once a .uf2 capable
bootloader is installed.

**Reply by robert-hh** [reply]
I prepared a change to the firmware for using the on-board flash for the file system, increasing the file system from 64kB to 1 MB and the code space from 184kB to 248kB. That extra space is used for more MicroPython features. The .uf2 is attached. On my brand new board it turned out that all sectors of the flash were write-locked after a power cycle. See the next post. 
[firmware.zip](https://github.com/user-attachments/files/17163372/firmware.zip)

**Reply by robert-hh** [reply]
There seems to be a problem with the board or flash chip. After a power cycle all sectors of the flash are write locked. I do not see any activity at the interface to the flash that's causing this. After running the unlock-script below the flash is unlocked and survives hard reboots until a power cycle.
@SirTobias Could you try if you have the same problem with your board?

```py
from machine import SPI, Pin, soft_reset
from time import sleep_ms

def wait(spi, cs):
    buf = bytearray(2)
    for i in range(1000):
        cs(0)
        spi.write_readinto(b"\x05\x00", buf)
        cs(1)
        if (buf[1] & 0x01) == 0:
            break;
        sleep_ms(1)
    return buf[1]

# instantiate SPI and CS objects
spi = SPI(5, mosi="PB22", miso="PB16", sck="PB23")
cs = Pin("PA13", Pin.OUT, value=1)

# get the initial status
status = wait(spi, cs)
print("Status: ", hex(status), "locked" if status & 0x0c else "unlocked")

if status & 0x0c:
# unlock all sectors
    cs(0)
    spi.write(b"\x06")
    cs(1)
    cs(0)
    spi.write(b"\x01\x00")
    cs(1)
    # get the final status
    status = wait(spi, cs)
    print("Status: ", hex(status), "locked" if status & 0x0c else "unlocked")
    soft_reset()
```
Edit: It turned out as well that the board cannot operate the flash with 24MHz SCK. So it's reduced to 12Mhz.

**Reply by robert-hh** [reply]
Just another variant of the above generic bootloader with ID's and boot label indicating the xplained pro board. Just text changes.
If you have already MicroPython running you can use the .uf2 variant uploaded though the target port. Then you have to reinstall MicroPython.

[bootloader-xplained-pro.zip](https://github.com/user-attachments/files/17174618/bootloader-xplained-pro.zip)

**Reply by robert-hh** [reply]
Finally I moved the unlock code into flash. It seems to be a standard behavior of that chip to have the flash locked after power up. Unlocking does not hurt if it is already unlocked.

[firmware_xplained_pro.zip](https://github.com/user-attachments/files/17178472/firmware_xplained_pro.zip)

**Reply by robert-hh** [reply, marked as answer]
This morning I replicated the steps for installing the bootloader again. I could not get .bin files installed by copying them onto the XPLAINED drive. Only Intel Hex files work. So there is still the step needed to convert the .bin file to an Intel HEX format.

[bootloader-xplained-pro-v3.16.0-15-gaa52b22.zip](https://github.com/user-attachments/files/17178808/bootloader-xplained-pro-v3.16.0-15-gaa52b22.zip)

**Comment by SirTobias** [comment]
Dear Robert, thank you very much for the extraordinary help. That worked out. 
I could also use your first bootloader answer (https://github.com/orgs/micropython/discussions/15886#discussioncomment-10771710)  the upload procedure for micropython and this worked out perfectly. Here, I could also use my self-compiled micropython for the samd21 xplained pro board. For the connection, I just have to connect via `picocom /dev/ttyACM0 -b 1200`. It worked out of the box to get the prompt ready.

**Reply by robert-hh** [reply]
Thank you for the feedback. I can now make a pull request. Note that you can use higher baudrates in the communication.  
The version of the firmware that I posted uses the external flash for the file system and offers more MicroPython features.
